client/driver/logging/rotator.go

```diff
@@ -70,28 +83,52 @@ func NewFileRotator(path string, baseFile string, maxFiles int,
 // equal to the maximum size the user has defined.
 func (f *FileRotator) Write(p []byte) (n int, err error) {
 	n = 0
-	var nw int
+	var forceRotate bool
 
 	for n < len(p) {
 		// Check if we still have space in the current file, otherwise close and
 		// open the next file
-		if f.currentWr >= f.FileSize {
+		if forceRotate || f.currentWr >= f.FileSize {
+			forceRotate = false
 			f.flushBuffer()
 			f.currentFile.Close()
 			if err := f.nextFile(); err != nil {
 				f.logger.Printf("[ERROR] driver.rotator: error creating next file: %v", err)
 				return 0, err
 			}
 		}
-		// Calculate the remaining size on this file
-		remainingSize := f.FileSize - f.currentWr
-
-		// Check if the number of bytes that we have to write is less than the
-		// remaining size of the file
-		if remainingSize < int64(len(p[n:])) {
-			// Write the number of bytes that we can write on the current file
-			li := int64(n) + remainingSize
-			nw, err = f.writeToBuffer(p[n:li])
+		// Calculate the remaining size on this file and how much we have left
+		// to write
+		remainingSpace := f.FileSize - f.currentWr
+		remainingToWrite := int64(len(p[n:]))
+
+		// Check if we are near the end of the file. If we are we attempt to
+		// avoid a log line being split between two files.
+		var nw int
+		if (remainingSpace - lineScanLimit) < remainingToWrite {
+			// Scan for new line and if the data up to new line fits in current
+			// file, write to buffer
+			idx := bytes.IndexByte(p[n:], newLineDelimiter)
+			if idx >= 0 && (remainingSpace-int64(idx)-1) >= 0 {
```

Comment:
Does this conditional work right if the first char in p is a new line, i.e idx=0? 
IIUC, It would write just the new line char to the buffer and then get back to the for loop. 

Might want to make a unit test for such edge cases though?

Reply:
Yeah that is what the minus 1 is for. Since a write of p[n:n] actually writes 1 byte so I have to check if there is room for whatever index + 1